File: 03_Spring/spring-mvc/src/main/java/com/springbook/biz/polymorphism/tv/LgTV.java
package com.springbook.biz.polymorphism.tv;

import com.springbook.biz.polymorphism.speaker.Speaker;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
public class LgTV implements TV {
    private Speaker speaker;
    private int price;

    public LgTV() {
        System.out.println("===> LgTV 객체 생성");
    }

    /*
    public LgTV(Speaker speaker) {
        this.speaker = speaker;
        System.out.println("===> LgTV 객체 생성");
    }

    public void setSpeaker(Speaker speaker) {
        System.out.println("===> setSpeaker() 호출");
        this.speaker = speaker;
    }

    public void setPrice(int price) {
        System.out.println("===> setPrice() 호출");
        this.price = price;
    }
    */

    @Override
    public void powerOn() {
        System.out.println("LgTV -- Turn on, price " + price);
    }

    @Override
    public void powerOff() {
        System.out.println("LgTV -- Turn off");
    }

    @Override
    public void volumeUp() {
        speaker.volumeUp();
    }

    @Override
    public void volumeDown() {
        speaker.volumeDown();
    }
}

File: 03_Spring/Ktn/src/main/java/com/example/polymorphism/tv/LgTV.kt
package com.example.kotlin_spring.polymorphism.tv

import com.example.kotlin_spring.polymorphism.speaker.Speaker
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Required
import org.springframework.stereotype.Component

@Component
class LgTV: TV {
    private lateinit var speaker: Speaker       // 필드 주입
    private var price = 2000000

    init {
        println("LgTV 객체 생성, price = $price")
    }

    fun setSpeaker(speaker: Speaker) {
        this.speaker = speaker
        println("LgTV -- setSpeaker 호출")
    }

    fun setPrice(price: Int) {
        this.price = price
        println("LgTV의 price -> $price")
    }

    override fun powerOn() {
        println("LgTV --- On")
    }

    override fun powerOff() {
        println("LgTV --- Off")
    }

    override fun volumeUp() {
        speaker.volumeUp()
    }

    override fun volumeDown() {
        speaker.volumeDown()
    }
}